Feature/rovecomm parse

Yajirushi12 opened the pull request.

thomas-gleiforst committed:
Merge pull request #97 from MissouriMRDT/dev

Merge 2020 dev branch into master (despite little testing...) to promote further development

Commit:
Fixed the packet slicing to work correctly on Javascript

Commit:
Changed message to be of type Buffer to match the nodejs imeplementation

Commit:
Prepared the code to start implementing RoveComm in JS

Commit:
Updated the UDP message event listener to call the parse function

Commit:
Added a textual description of the packing method in the parse function to help my memory

Commit:
Implemented decode, still needs testing

Commit:
Finished implementing and fixing the parse and decodePacket functions for non-string variables

Commit:
Restored the UDP port to 8080

Commit:
Merge branch 'master' of https://github.com/MissouriMRDT/BaseStation_Software into feature/rovecomm_parse

Commit:
Oops, put UDP back to 8081 correctly

Yajirushi12 commented:
Going to add String decoding in a near future PR, wanted to get the fully functional number decoding into dev asap to support development by others.

Commit:
Removed the emitHelper function because I had to disable the eslint anyway

Yajirushi12 commented on Electron/Core/RoveProtocol/Rovecomm.ts:
```diff
@@ -4,6 +4,87 @@ const dgram = require("dgram")
+  switch (size) {
+    case 1:
+      while (i < dataLength) {
+        retArray.push(Number(data[i]))
```
We probably need to consider what exactly to return as. Number, BigInt, and array of uint/int 8/16/32?

Yajirushi12 replied:
I think standardizing on either Number or BigInt is probably the best way to go to provide a sufficiently abstracted interface for other components and not require them to know which data type they're receiving.

Thomas-McKanna committed:
Added more precise RoveComm packet format in comment.

Thomas-McKanna committed:
Simplified parse and decodePacket.

Yajirushi12 commented on Electron/Core/RoveProtocol/Rovecomm.ts:
```diff
@@ -7,49 +7,41 @@ const EventEmitter = require("events")
-        retArray.push(Number(data[i]))
-        i += 1
+      readBytes = function (i: number) {
+        return data.readUInt8(i)
```
I don't think using a readUint or readInt function here is the correct course of action, We have some packets which contain signed information versus some that contain unsigned information under last years standards iirc.

Commit:
Changed the emit back to DATAID[dataId] that I messed up during an earlier commit

Thomas-McKanna committed:
decodePacket now uses data type to read bytes.

Thomas-McKanna committed:
Merge branch 'feature/rovecomm_parse' of github.com:MissouriMRDT/BaseStation_Software into feature/rovecomm_parse

Yajirushi12 commented on Electron/Core/RoveProtocol/Rovecomm.ts:
```diff
@@ -4,40 +4,62 @@ const dgram = require("dgram")
-      }
+  let readBytes: (i: number) => number
+  let size: number
+  switch (dataType) {
```
Shouldn't this switch on DataTypes[dataType]?

thomas-gleiforst replied:
I don't believe so.. At least not the way we have the rest of the switch set up currently. From Rovecomm Manifest, dataType is the number from the enum, such as 6 representing float. So we should switch on the number 6, and figure out which enum in DataTypes = 6. Correct?

rkbhkp committed:
Added Conosle

Added the conssole file and everything that makes it work, still needs to be able to update

thomas-gleiforst committed:
Update ports to match old basestation

thomas-gleiforst committed:
Merge remote-tracking branch 'origin/feature/Consol' into feature/rovecomm_parse

thomas-gleiforst committed:
Added console all packets print to

thomas-gleiforst committed:
Manifest update (+implementation)

thomas-gleiforst committed:
Removed old data id structure

(and updated a discrepancy with spectrometer)

thomas-gleiforst committed:
Added size list in place of inline

thomas-gleiforst committed:
No longer export parse

Parse should really only be used inside of rovecomm

thomas-gleiforst committed:
Made TCP listen more similar to UDP

thomas-gleiforst committed:
Implement sendCommand and Udp send

thomas-gleiforst committed:
Updated implementation example

thomas-gleiforst commented:
Note: This PR has expanded beyond just "rovecomm parse" and should now be all rovecomm functionality other than outgoing TCP connections.

thomas-gleiforst committed:
Fixed improper reference to old dataIds

thomas-gleiforst committed:
update DataLength to dataSizes

thomas-gleiforst committed:
Clean up console logs on recieve

While this may be helpful since we are still in debugging, we don't want to miss this and affect performance. Plus, everything is being logged to our in-house console.

thomas-gleiforst committed:
Removed "any" variables

thomas-gleiforst committed:
In must be called on the object, not on a list

Fixed this implementation in other places, missed here.

thomas-gleiforst committed:
Add thorough documentation

Also fixed odd default port value

thomas-gleiforst committed:
Remove deprecated Packet.ts

thomas-gleiforst committed:
Instead of parseInt(port), just make ports ints

thomas-gleiforst committed:
Move DataTypes and dataSizes to manifest

Slightly more consistent to keep these changes here, and allows the manifest to use them.

thomas-gleiforst committed:
Added comments to the manifest

thomas-gleiforst committed:
Switch dataType out for the actual enum name

Also add comments for drive board

thomas-gleiforst committed:
Remove unnecessary dependency

EliVerbrugge commented on Electron/Core/RoveProtocol/Rovecomm.ts:
```diff
@@ -1,85 +1,273 @@
+        dataId = DATAID[i].Commands[dataIdStr].dataId
+        break
+      }
+    }
```
Are you handling a packet that cannot be found in the manifest? Right now it looks like you might continue with incomplete data.

thomas-gleiforst replied:
We were not. This has been fixed.

EliVerbrugge commented on Electron/Core/RoveProtocol/Rovecomm.ts:
```diff
@@ -1,85 +1,273 @@
+   *   0                   1                   2                   3
+   *   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+   *  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+   *  |    Version    |            Data Id            |  Data Length  |
```
dataLength may be a little ambiguous, it's technically the data count. Length might imply that its the size in bytes, while it's the number of seperate data entries. Not a big deal either way.

Yajirushi12 replied:
Yeah, we should probably switch this to Data Count in the two current instances to ensure easier readability for future maintainers.

EliVerbrugge commented on Electron/Core/RoveProtocol/Rovecomm.ts:
```diff
@@ -1,85 +1,273 @@
+
+    // First emit is for the dataIdString (like DriveSpeeds)
+    // eslint-disable-next-line @typescript-eslint/no-use-before-define
+    rovecomm.emit(dataIdStr, data)
```
Do you want to handle telemetry we can't find in the manifest? Or is a null dataIdStr event good enough?

Yajirushi12 replied:
With the current structure, the emit with a specific string will be used for easy consumption of messages by down-stream components, so leaving it as null will just prevent any component from accessing the data. It will still be logged through the "all" emit just after.

It would probably be best to have an additional warning in the event that a "null" emit occurs to provide real-time feedback as to packet definitions varying between the board implementation and the basestation implementation though.

thomas-gleiforst replied:
Emitting this null event will log the data and the dataId to the Console. The hope is that this will be enough information for someone at basestation to diagnose this issue and fix the board code or add the dataId to the manifest.

Yajirushi12 commented:
I'm probably going to close this PR without merging soon, I was intending for it to be merged after parsing was completed but it looks like we're going to be putting all of the rovecomm implementation in this branch before merging.

thomas-gleiforst committed:
Add resubscribe

thomas-gleiforst committed:
Added event emitter when invalid data sent

thomas-gleiforst committed:
Change dataLength to DataCount

thomas-gleiforst commented:
> I'm probably going to close this PR without merging soon, I was intending for it to be merged after parsing was completed but it looks like we're going to be putting all of the rovecomm implementation in this branch before merging.

While the name of the PR is no longer accurate, we do still plan to merge these features today. Please leave open.

Commit:
Fixed the dataIdStr grabbing in parse()
